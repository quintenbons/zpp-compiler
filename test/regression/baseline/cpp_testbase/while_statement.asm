section .data

section .rodata

section .bss

section .text
	extern printnum:function

	global main:function
	global _start:function

_start:
	call main
	mov rdi, rax            ; Exit code (0) expects return of main to be put in rax for now
	mov rax, 60                  ; Syscall number for exit (60)
	syscall                      ; Make the syscall

main:
	push rbp                 ; Save the base pointer
	mov rbp, rsp              ; Set base pointer to current stack pointer
	sub rsp, 4 ; Creating space on the stack
	mov rax, 5 ; Loading number literal
	mov [rbp-4], eax ; Storing value in memory
._U0_while:
	mov eax, [rbp-4] ; Loading value from memory
	test rax,rax
	jz ._U1_while.end
	mov edi, [rbp-4] ; Loading value from memory
	call printnum
	mov eax, [rbp-4] ; Loading value from memory
	mov rbx, 1 ; Loading number literal
	sub rax, rbx
	mov [rbp-4], rax ; Storing value in memory
	jmp ._U0_while
._U1_while.end:
	mov rax, 0 ; Loading number literal
	mov rsp, rbp              ; Restoring stack pointer
	pop rbp                   ; Restore the base pointer
	ret
