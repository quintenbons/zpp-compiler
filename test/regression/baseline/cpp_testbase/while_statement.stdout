/home/qbons/zpp-compiler/cpp_testbase/while_statement.cpp
== Parsing

== Done parsing
[35m[Node_TranslationUnit] Function count: 1
[0m[35m  [Node_Function] ReturnType: int ; Name: main ; ParamCount: 0
[0m[35m    [Node_Type] int
[0m[35m    [Node_FunctionParameterList] ParameterCount: 0
[0m[35m    [Node_CodeBlock] InstructionCount: 3
[0m[35m      [Node_Declaration] Type: int ; Assignment: 1
[0m[35m        [Node_Variable] i
[0m[35m        [Node_NumberLiteral] 5
[0m[35m      [Node_WhileStatement] 
[0m[35m        [Node_Variable] i
[0m[35m        [Node_CodeBlock] InstructionCount: 2
[0m[35m          [Node_FunctionCall] printnum ; ArgumentCount: 1
[0m[35m            [Node_Variable] i
[0m[35m          [Node_Assign] 
[0m[35m            [Node_Variable] i
[0m[35m            [Node_BinaryOperation] -
[0m[35m              [Node_Variable] i
[0m[35m              [Node_NumberLiteral] 1
[0m[35m      [Node_ReturnStatement] 
[0m[35m      [Node_NumberLiteral] 0
[0m[35m[Node_TranslationUnit] Class count: 0
[0m
== Decorating
[35mScopeStack of 3 scopes
[0m[35m[Scope] id=0 ; parent=0
  [Type] id=2 ; name=char ; size=1
  [Type] id=2 ; name=int ; size=4
  [Type] id=0 ; name=void ; size=0
  [Function] id=2 name=main ; returnType=int ; parameters=
  [Function] id=1 name=printnum ; returnType=void ; parameters=int 
[0m[35m[Scope] id=1 ; parent=0
  [Variable] id=1 ; name=i ; location=LocalStackOffset: byteSize=4, byteOffset=4
[0m[35m[Scope] id=2 ; parent=1
[0m
== Done decorating
[35m[Node_TranslationUnit] Function count: 1
[0m[35m  [Node_Function] ReturnType: int ; Name: main ; ParamCount: 0
[0m[35m    [Node_Type] int
[0m[34m      [Decoration] TypeDescription: int ; Size: 4 ; Id: 2
[0m[35m    [Node_FunctionParameterList] ParameterCount: 0
[0m[35m    [Node_CodeBlock] InstructionCount: 3
[0m[35m      [Node_Declaration] Type: int ; Assignment: 1
[0m[35m        [Node_Variable] i
[0m[34m          [Decoration] VariableDescription: Id: 1
[0m[35m        [Node_NumberLiteral] 5
[0m[35m      [Node_WhileStatement] 
[0m[35m        [Node_Variable] i
[0m[34m          [Decoration] VariableDescription: Id: 1
[0m[35m        [Node_CodeBlock] InstructionCount: 2
[0m[35m          [Node_FunctionCall] printnum ; ArgumentCount: 1
[0m[35m            [Node_Variable] i
[0m[34m              [Decoration] VariableDescription: Id: 1
[0m[35m          [Node_Assign] 
[0m[35m            [Node_Variable] i
[0m[34m              [Decoration] VariableDescription: Id: 1
[0m[35m            [Node_BinaryOperation] -
[0m[35m              [Node_Variable] i
[0m[34m                [Decoration] VariableDescription: Id: 1
[0m[35m              [Node_NumberLiteral] 1
[0m[35m      [Node_ReturnStatement] 
[0m[35m      [Node_NumberLiteral] 0
[0m[34m    [Decoration] FunctionDescription: main ; Id: 2
[0m[35m[Node_TranslationUnit] Class count: 0
[0m
== Generating code
== Generated asm to a.asm:
section .data

section .rodata

section .bss

section .text
	extern printnum:function

	global main:function
	global _start:function

_start:
	call main
	mov rdi, rax            ; Exit code (0) expects return of main to be put in rax for now
	mov rax, 60                  ; Syscall number for exit (60)
	syscall                      ; Make the syscall

main:
	push rbp                 ; Save the base pointer
	mov rbp, rsp              ; Set base pointer to current stack pointer
	sub rsp, 4 ; Creating space on the stack
	mov rax, 5 ; Loading number literal
	mov [rbp-4], eax ; Storing value in memory
._U0_while:
	mov eax, [rbp-4] ; Loading value from memory
	test rax,rax
	jz ._U1_while.end
	mov edi, [rbp-4] ; Loading value from memory
	call printnum
	mov eax, [rbp-4] ; Loading value from memory
	mov rbx, 1 ; Loading number literal
	sub rax, rbx
	mov [rbp-4], rax ; Storing value in memory
	jmp ._U0_while
._U1_while.end:
	mov rax, 0 ; Loading number literal
	mov rsp, rbp              ; Restoring stack pointer
	pop rbp                   ; Restore the base pointer
	ret
== Generated .o as a.o:
== Generating exe as a.out:
